### DISCUSSION TITLE
I don't understand the '/' syntax for base_url

### DISCUSSION THREAD

**Discussion by JulesArcher** [fonsp/Pluto.jl#2558] [Questions about using Pluto] [answered]
Hi,

I am trying to connect to a Pluto service that includes the command to run Pluto: the problem seems to be when trying to specify the base_url it fails to launch the service (when I connect to port 8000 only via http, I can get a connection):

Pluto.run(threads=6, port=8000,launch_browser=false, require_secret_for_access=false,base_url="https://example.net/")

The Configuration.jl file says "it should start be a valid path starting and ending with a '/'". This statement is not clear to me, and  I don't know how the back-slash should be used. Can someone please provide me with the right syntax for how the base_url should be specified?

Thanks!

**Comment by disberd** [comment]
I think Pluto does not work out of the box with HTTPS, if I remember correctly @fonsp mentioned to me once that you can't simply open a Pluto server on a remote that is accessible through https and have it work via HTTPS, there is some additional things you'd have to set up on the server side to make this work.
Maybe also @pankgeorg can shed some light into this

**Reply by pankgeorg** [reply]
Exactly - you need to either handle SSL (e.g. using MbedTLS) or have a reverse proxy that handles that for you (which is the recommended setup).

When setting up the reverse proxy, you may want to have multiple services on the same server. You have two options there: either have a new server block for Pluto with a subdomain, like pluto.example.com, or forward a subpath like example.com/pluto. If you go with scenario 2, you need to tell Pluto about it, because all paths that arrive to Pluto will be prepended by '/pluto/' and therefore not match by default. So, we add the base_url and fix that. Note that you may rewrite the proxied path and remove the base url on the reverse proxy - in that case you don't need base_url.

**Reply by JulesArcher** [reply]
Thanks @disberd and @pankgeorg for your replies, here's some additional follow-up that helps to answer my question but leads to new issues:

When I run the following below:

Pluto.run(require_secret_for_access=false,launch_browser=false,base_url="https://mypluto.net/")

I get AssertionError: base_url "https://mypluto.net/" should start with a '/'

Now when I run:  Pluto.run(require_secret_for_access=false,launch_browser=false,base_url="/https://mypluto.net/")

the Pluto server launches, which solves the '/' format issue, but now the info says:

Go to http://localhost:1234/https://mypluto.net/ in your browser to start writing ~ have fun!

How do I over-ride the localhost:1234 and just have  https://mypluto.net/ as the location?

Thanks!

**Reply by pankgeorg** [reply]
`base_url` is not the option you need; you need `host="0.0.0.0"`. Can you confirm if it worked?

**Reply by JulesArcher** [reply]
Hi @pankgeorg,

Thanks again for your quick reply!

Progress! I had to also include the url to my https registered domain specified in the root_url option as well as the specific port that the proxy server is specified as follows:
 
Pluto.run(require_secret_for_access=false,port=8000,launch_browser=false,host="0.0.0.0",root_url="https://mypluto.net/")

The Pluto server starts up, and when I direct my browser to my https domain, I can now see the Pluto start-up page "Welcome to Pluto.jl", but cannot create a new Pluto file, or access the directory structure on my system to open an existing Pluto notebook.

Any ideas why I can't create or access Pluto notebooks? Or is there possibly something still not right with my Pluto.run options?

Thanks!

**Reply by pankgeorg** [reply]
How does your HTTPS work? Is it handled by some managed service like Cloudflare? Or are you doing let'sencrypt certificates? Are you using PlutoSliderServer? Do you have a reverse proxy? If yes, where is it pointing/how does your location block look like?

Deployment is generally hard, don't quit even if you feel overwhelmed! It will work soon!!

**Comment by fonsp** [comment]
It sounds like you should not set `base_url` for your application, just leave the setting at the default and it will work. You don't need a setting to run your server on a TLD (example.com), this is something you configure with your registrar and DNS. Run on 0.0.0.0:80 as @pankgeorg suggested.

**Reply by fonsp** [reply]
I think the purpose of `base_url` is to run your server at `http://localhost:1234/some/custom/subpath/` instead of `http://localhost:1234/`. (with `base_url="/some/custom/subpath/"`)

**Reply by JulesArcher** [reply]
Hi @fonsp and @pankgeorg,

Thank you for your helpful replies and kind words of encouragement to try to keep going with this!
A few more details:
1) Yes, I have set up my own linux VPS with nginx server with reverse proxy enabled, and a registered domain name, with https secured by letsencrypt certificate. 
2) I have previously set up a PlutoSliderServer running as a service, and it's been running great for over a year now! I can access this PlutoSliderServer using my registered domain name via https.

3) The command that successfully launches the PlutoSliderServer is as follows:
PlutoSliderServer.run_directory("home/hpcuser/notebooks", SliderServer_host="0.0.0.0",Export_slider_server_url="https://myserver.net")

4) Instead of a PlutoSliderServer, I am trying to launch a Pluto server. The run command that gets me close is the following:
Pluto.run(Pluto.ServerSession(secret="hello",options=Pluto.Configuration.from_flat_kwargs(host="0.0.0.0", port=8000,launch_browser=false)))

(more complicated command is to allow a pre-defined secret to be used)

As I mentioned, "the welcome to Pluto.jl" main page is served when I try to connect via https:myserver.net, however the page is only "static", i.e. I can't load any existing notebooks or create a new one.

Any ideas on why the Pluto welcome page is served as a static one only?

Thanks!

**Reply by JulesArcher** [reply]
Sorry, a correction to to PlutoSliderServer run: it also needs to have the port specified:
PlutoSliderServer.run_directory("home/hpcuser/notebooks", SliderServer_host="0.0.0.0",SliderServer_port=8000, Export_slider_server_url="https://myserver.net/")

**Reply by fonsp** [reply, marked as answer]
Interesting! It sounds like HTTP traffic is working, but WebSocket traffic is not. You can verify this the browser's devtools, check the Console for error messages, or the Network panel for a failed WS request.

You probably need to enable websockets in your nginx proxy: this is something that Pluto uses, but PlutoSliderServer does not (this was one of the reasons, yay!)

**Reply by JulesArcher** [reply]
@fonsp, thank you for the reply and suggestions! Your assessment of the issue and possible resolution makes perfect sense.
As I am able to create an ssh tunnel to my VPS with Pluto running as a service, I think this will satisfy my computational needs for now, and trying to resolve this is not urgent.
This thread has wandered away from my original questions, but I've learned a lot (as one typically does when things go wrong)!
Thanks again for your and @pankgeorg 's help with this. I will mark this original discussion as answered, but will re-visit with a new discussion if necessary.

Thanks!

**Comment by pankgeorg** [comment]
Very welcome!

For future reference, here is a working nginx location block for HTTP/WS forwarding

```nginx
        location /pluto { # if you keep "/pluto" here you also do `base_url` = `/pluto/`. Otherwise do `/` (also depends on what else you have here)
			# Not really needed but good practice
            proxy_set_header        Host $host;
            proxy_set_header        X-Real-IP $remote_addr;
            proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header        X-Forwarded-Proto $scheme;

			# Timeout at 5 minutes; if you upload files, do more
            proxy_read_timeout      300s;

            proxy_pass              http://localhost:port/;

            # WebSocket support
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
```
